### PR TITLE
Update team

### PR DESCRIPTION
**Pull request by drouba**

### CODE REVIEW

File: data/init.sql [hunk 1]
```diff
@@ -19,7 +19,7 @@ CREATE TABLE IF NOT EXISTS users (
 	student_id INT UNSIGNED,
 	created_at TIMESTAMP,
 	updated_at TIMESTAMP,
-	UNIQUE (email)
+	UNIQUE (email, student_id)
 );
```
Comment (jamestjw) [inline data/init.sql:23]: ok wait a sec does this mean that each pair of email and studentID has to be unique? i.e. its ok to have the same student ID as long as the email is different and vice versa?

File: data/init.sql [hunk 2]
```diff
@@ -251,3 +251,5 @@ CREATE TABLE IF NOT EXISTS meetings (
 	FOREIGN KEY (team_id) REFERENCES teams(id) 
 );
 
+
+
```
Comment (jamestjw) [inline data/init.sql:255]: 💀

File: modules/models/team_member.php [hunk 3]
```diff
@@ -14,9 +14,13 @@ class TeamMember extends Record
         "teams"=> array(
             "class_name" => "Team",
             "foreign_key" => "team_id"
-            )
+        ),
+        "lecture" => array(
+            "class_name" => "lecture",
+            "foreign_key" => "lecture_id"
+        )
```
Comment (jamestjw) [inline modules/models/team_member.php:21]: How does this work? This table doesn't have this column

File: public/course_lecture.php [hunk 4]
```diff
@@ -157,7 +158,7 @@
     <?php }
 ?> 
 
-
+<?php if(get_current_role() == 'instructor') { ?>
 <div class="container">
```
Comment (jamestjw) [inline public/course_lecture.php:162]: Should we also check that the current user actually teaches the course?

File: public/create_team.php [hunk 5]
```diff
@@ -1,22 +1,36 @@
 <?php require_once(dirname(__FILE__)."/../modules/ensure_logged_in.php"); ?>
 <?php include "templates/header.php"; ?>
 
-<?php if(get_current_role() == 'instructor') {
+<?php 
+$lecture_id = $_GET['id'];
+if(get_current_role() == 'instructor') {
 
 require_once "../modules/models/user.php";
```
Comment (jamestjw) [inline public/create_team.php:8]: Probably worth ensuring that the user is actually teaching the course

File: public/update_team.php [hunk 6]
```diff
@@ -0,0 +1,152 @@
+<?php require_once(dirname(__FILE__)."/../modules/ensure_logged_in.php"); ?>
+<?php include "templates/header.php"; ?>
+
+<?php 
+$lecture_id = $_GET['id'];
+
+if(get_current_role() == 'instructor') {
+
```
Comment (jamestjw) [inline public/update_team.php:8]: instructor permissions again

File: public/update_team.php [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+if (isset($_POST["submit"])) {
+
+    $team_to_edit_id = $_POST['team'];
+    if (isset($_POST['team_member_delete'])){
+        $students_to_remove = $_POST['team_member_delete'];
+    }
+    else {
+        $students_to_remove = null;
+    }
+
```
Comment (jamestjw) [inline public/update_team.php:68]: `$students_to_remove = $_POST['team_member_delete'] ?? null;` should work instead of the entire if else statement

Comment (jamestjw) [reply public/update_team.php:68]: or maybe even $students_to_remove = $_POST['team_member_delete'] ?? [];

this way you can skip the null check below

File: public/update_team.php [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    }
+    else {
+        $students_to_remove = null;
+    }
+    
+
+    if(!is_null($students_to_remove)){ 
+        foreach($students_to_remove as $remove){
+            $delete = TeamMember::find_by(array('user_id' => $remove, 'team_id' => $team_to_edit_id ));
+            $delete->delete();
```
Comment (jamestjw) [inline public/update_team.php:73]: $delete is nullable here, this could give an error

File: public/update_team.php [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        $team_member->team_id = $team_to_edit_id;
+        $team_member->user_id = $_POST['student_selection'];
+
+        try {
+            $team_member->save();
+        } catch (PDOException $error) {
+            echo "<br>" . $error->getMessage();
+        }
+    }
+    echo "<h5>Team updated succesfully </h5>";
```
Comment (jamestjw) [inline public/update_team.php:93]: is this even visible?

File: public/update_team.php [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+                    <td><?php echo escape($student->user->email); ?></td>
+                    <td><?php echo escape($student->user->created_at);  ?> </td>
+                    <td><input class="single-checkbox" type="checkbox" id="<?php echo $student->user_id; ?>" name="team_member_delete[]" value="<?php echo $student->user_id; ?>"> Remove from team</td>
+                <?php }?>
+                </tbody>
+            </table>
+            <?php if (count($students_in_team)<4){ ?>
+                Add Student: 
+                <select Name="student_selection" id="student_selection">
+                    <option value="">----Select----</option>
```
Comment (jamestjw) [inline public/update_team.php:132]: this should be disabled or something, people shouldnt be allowed to choose this

File: public/uploadStudentList.php [hunk 11]
```diff
@@ -14,39 +14,66 @@
     echo "<br>" . $error->getMessage();
 }
 
+try {
+    $all_students = User::where(array('is_instructor' => '0', 'is_admin' => '0'));
```
Comment (jamestjw) [inline public/uploadStudentList.php:18]: should we instead check that student_id != null?

File: public/uploadStudentList.php [hunk 12]
```diff
@@ -14,39 +14,66 @@
...
-            $user->is_ta = 0;
-            $user->password_digest = password_hash("welcome", PASSWORD_DEFAULT);
-
-            try {
-                $user->save();
+
+            $existing_student = array_search($studentData[0], array_column($all_students, 'student_id'));
+            $student_section_exists = in_array($all_students[$existing_student]->id, array_column($all_student_in_section, 'user_id'));
+
+            if($existing_student === false){
```
Comment (jamestjw) [inline public/uploadStudentList.php:36]: if(!$existing_student) ?

**Review by jamestjw — CHANGES_REQUESTED** [review]

**Review by chris7neves — APPROVED** [review]

File: public/update_team.php (hunk 10 above)
Comment (jamestjw) [reply public/update_team.php:137]: yeah  zack did it that way its not good

Comment (jamestjw) [reply public/update_team.php:137]: you can't send the form without choosing anything

File: modules/models/record.php [hunk 13]
```diff
@@ -169,7 +169,9 @@ public function save()
         $conn = getConnection();
         execute_sql_query($sql, $new_obj, $conn);
 
-        $this->id = $conn->lastInsertId();
+        if (in_array("id", get_called_class()::getAttrs())) {
+            $this->id = $conn->lastInsertId();
+        }
         $this->is_new_record = false;
```
[the remark this answers is not in the document]
Comment (jamestjw) [reply modules/models/record.php:175]: this was the fix

**Review by jamestjw — APPROVED** [review]

**Pull request merged**